File: armyknife/internal/rails_message_encryptor/rails_message_encryptor.go
package rails_message_encryptor

import (
	"armyknife/internal/marshal/ruby"
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"strings"

	"golang.org/x/xerrors"
)

const (
	AuthTagLength = 16
	Separator     = "--"
)

type Cipher string

const (
	AESGCM Cipher = "AES-GCM"
)

type RailsMessageEncryptor struct {
	secret []byte
	cipher Cipher
}

type Options struct {
	Cipher Cipher
}

func NewRailsMessageEncryptor(secret string, options *Options) *RailsMessageEncryptor {
	c := AESGCM
	if options != nil {
		if options.Cipher != "" {
			c = options.Cipher
		}
	}

	return &RailsMessageEncryptor{
		secret: []byte(secret),
		cipher: c,
	}
}

func (m *RailsMessageEncryptor) Encrypt(value string) (*string, error) {
	switch m.cipher {
	case AESGCM:
		decodedSecret := make([]byte, hex.DecodedLen(len(m.secret)))
		if _, err := hex.Decode(decodedSecret, m.secret); err != nil {
			return nil, xerrors.Errorf("failed to decode hex secret: %w", err)
		}
		block, err := aes.NewCipher(decodedSecret)
		if err != nil {
			return nil, xerrors.Errorf("failed to create new cipher: %w", err)
		}

		gcm, err := cipher.NewGCM(block)
		if err != nil {
			return nil, xerrors.Errorf("failed to create new GCM: %w", err)
		}

		iv := make([]byte, gcm.NonceSize())
		if _, err := rand.Read(iv); err != nil {
			return nil, xerrors.Errorf("failed to generate random iv: %w", err)
		}

		serialized, err := ruby.Dump(value)
		if err != nil {
			return nil, xerrors.Errorf("failed to serialize value: %w", err)
		}
		result := gcm.Seal(nil, iv, []byte(serialized), nil)
		l := len(result)
		encryptedData := result[:l-AuthTagLength]
		tag := result[l-AuthTagLength:]

		s := strings.Join([]string{
			base64.StdEncoding.EncodeToString(encryptedData),
			base64.StdEncoding.EncodeToString(iv),
			base64.StdEncoding.EncodeToString(tag),
		}, Separator)
		return &s, nil

	default:
		return nil, errors.New("unsupported cipher")
	}
}

func (m *RailsMessageEncryptor) Decrypt(message string) (*string, error) {
	switch m.cipher {
	case AESGCM:
		parts := strings.Split(message, Separator)
		if len(parts) != 3 {
			return nil, errors.New("invalid message format")
		}

		data, err := base64.StdEncoding.DecodeString(parts[0])
		if err != nil {
			return nil, xerrors.Errorf("failed to decode base64 data: %w", err)
		}

		iv, err := base64.StdEncoding.DecodeString(parts[1])
		if err != nil {
			return nil, xerrors.Errorf("failed to decode base64 iv: %w", err)
		}

		tag, err := base64.StdEncoding.DecodeString(parts[2])
		if err != nil {
			return nil, xerrors.Errorf("failed to decode base64 tag: %w", err)
		}

		decodedSecret := make([]byte, hex.DecodedLen(len(m.secret)))
		if _, err := hex.Decode(decodedSecret, m.secret); err != nil {
			return nil, xerrors.Errorf("failed to decode hex secret: %w", err)
		}
		block, err := aes.NewCipher(decodedSecret)
		if err != nil {
			return nil, xerrors.Errorf("failed to create new cipher: %w", err)
		}

		gcm, err := cipher.NewGCM(block)
		if err != nil {
			return nil, xerrors.Errorf("failed to create new GCM: %w", err)
		}

		result, err := gcm.Open(nil, iv, append(data, tag...), nil)
		if err != nil {
			return nil, xerrors.Errorf("failed to decrypt data: %w", err)
		}

		deserialized, err := ruby.Load(string(result))
		if err != nil {
			return nil, xerrors.Errorf("failed to deserialize data: %w", err)
		}
		s := deserialized.(string)
		return &s, nil
	default:
		return nil, errors.New("unsupported cipher")
	}
}

File: google-app-scripts/sheethub/src/rails_message_encryptor.ts
import {dump, load} from "./marshal/ruby";
// @ts-ignore
import sjcl from "sjcl";

const AUTH_TAG_LENGTH = 16 * 8;
const SEPARATOR = "--";

enum Cipher {
    AES_GCM = "AES-GCM",
}

class RailsMessageEncryptor {
    private readonly secret: string;
    private readonly cipher: Cipher;

    constructor(secret: string, options: { cipher?: Cipher } = {}) {
        this.secret = secret;
        this.cipher = options.cipher || Cipher.AES_GCM;
    }

    public addEntropy(size: number): void {
        if (typeof window !== 'undefined' && window.crypto) {
            sjcl.random.addEntropy(new Uint32Array(32), size, "crypto.getRandomValues");
        } else if (typeof global !== 'undefined' && global.crypto) {
            sjcl.random.addEntropy(new Uint32Array(32), size, "crypto.getRandomValues");
        } else {
            sjcl.random.addEntropy(new Uint32Array(32), size);
            console.warn("No secure random number generator available.");
        }
    }

    public encrypt(value: string): string {
        switch (this.cipher) {
            case Cipher.AES_GCM:
                const iv = sjcl.random.randomWords(3); // 12
                const result = sjcl.mode.gcm.encrypt(
                    new sjcl.cipher.aes(sjcl.codec.hex.toBits(this.secret)),
                    sjcl.codec.utf8String.toBits(dump(value)),
                    iv,
                    [],
                    AUTH_TAG_LENGTH,
                );
                const l = sjcl.bitArray.bitLength(result);
                const tag = sjcl.bitArray.bitSlice(result, l - AUTH_TAG_LENGTH);
                const data = sjcl.bitArray.bitSlice(result, 0, l - AUTH_TAG_LENGTH);

                return [
                    sjcl.codec.base64.fromBits(data),
                    sjcl.codec.base64.fromBits(iv),
                    sjcl.codec.base64.fromBits(tag)
                ].join(SEPARATOR);
            default:
                throw new Error(`Unsupported cipher: ${this.cipher}`);
        }
    }

    public decrypt(message: string): string {
        switch (this.cipher) {
            case Cipher.AES_GCM:
                const [data, iv, tag] = message.split(SEPARATOR);
                if (!data || !iv || !tag) {
                    throw new Error("Invalid message format.");
                }

                const result = sjcl.codec.utf8String.fromBits(sjcl.mode.gcm.decrypt(
                    new sjcl.cipher.aes(sjcl.codec.hex.toBits(this.secret)),
                    sjcl.bitArray.concat(sjcl.codec.base64.toBits(data), sjcl.codec.base64.toBits(tag)),
                    sjcl.codec.base64.toBits(iv),
                    [],
                    AUTH_TAG_LENGTH,
                ));

                return load(result) as string;
            default:
                throw new Error(`Unsupported cipher: ${this.cipher}`);
        }
    }
}

export {RailsMessageEncryptor, Cipher};
